openfe/setup/ligand_network_planning.py

```diff
@@ -72,18 +76,43 @@ def generate_radial_network(ligands: Iterable[SmallMoleculeComponent],
 
 
 def minimal_spanning_graph(ligands: Iterable[SmallMoleculeComponent],
-                           mappers: Iterable[LigandAtomMapper], scorer=None):
+                           mappers: Iterable[LigandAtomMapper],
+                           scorer: Callable[[LigandAtomMapping], float]):
     """Plan a Network which connects all ligands with minimal cost
 
     Parameters
     ----------
-    ligands : Iterable of rdkit Molecules
+    ligands : Iterable[SmallMoleculeComponent]
       the ligands to include in the Network
-    mappers : Iterable of AtomMappers
+    mappers : Iterable[LigandAtomMapper]
       the AtomMappers to use to propose mappings.  At least 1 required,
       but many can be given, in which case all will be tried to find the
       lowest score edges
     scorer : Scoring function
-      any callable which takes an LigandAtomMapping and returns a float
+      any callable which takes a LigandAtomMapping and returns a float
     """
-    raise NotImplementedError
+    nodes = list(ligands)
+
+    # First create a network with all the proposed mappings (scored)
+    mapping_generator = itertools.chain.from_iterable(
+        mapper.suggest_mappings(molA, molB)
+        for molA, molB in itertools.combinations(nodes, 2)
+        for mapper in mappers
+    )
+    mappings = [mapping.with_annotations({'score': scorer(mapping)})
+                for mapping in mapping_generator]
+    network = Network(mappings, nodes=nodes)
+
+    # Next analyze that network to create minimal spanning network. Because
+    # we carry the original (directed) LigandAtomMapping, we don't lose
+    # direction information when converting to an undirected graph.
+    min_edges = nx.minimum_spanning_edges(nx.MultiGraph(network.graph),
```

Comment:
does nx silently provide a non spanning network if there were no edges for a particular node?

Reply:
Yes. Test added in https://github.com/OpenFreeEnergy/openfe/pull/139/commits/a4a30e3c5dc7730033e17066674ef89f256b544f shows that it would happen -- we catch the error I added, not something from nx.

Maybe this is a difference between `nx.minimum_spanning_edges` vs. some of the other related nx functions?